v2/pkg/subscraping/sources/dnsdb/dnsdb.go

```diff
@@ -45,38 +65,99 @@ func (s *Source) Run(ctx context.Context, domain string, session *subscraping.Se
-			if line == "" {
-				continue
+		var offsetMax uint64
+		// if the rate.OffsetMax == rateNA then the ?offset= query parameter is not allowed
+		if rate.OffsetMax.String() != rateNA {
+			offsetMax, err = strconv.ParseUint(rate.OffsetMax.String(), 10, 64)
+			if err != nil {
+				results <- subscraping.Result{Source: sourceName, Type: subscraping.Error, Error: err}
+				s.errors++
+				return
 			}
-			var response dnsdbResponse
-			err = jsoniter.NewDecoder(bytes.NewBufferString(line)).Decode(&response)
+		}
+
+		var respCond string
+		for respCond == "" {
+			url := urlTemplate + queryParams.Encode()
+
+			resp, err := session.Get(ctx, url, "", headers)
 			if err != nil {
-				results <- subscraping.Result{Source: s.Name(), Type: subscraping.Error, Error: err}
+				results <- subscraping.Result{Source: sourceName, Type: subscraping.Error, Error: err}
 				s.errors++
+				session.DiscardHTTPResponse(resp)
 				return
 			}
-			results <- subscraping.Result{
-				Source: s.Name(), Type: subscraping.Subdomain, Value: strings.TrimSuffix(response.Name, "."),
+
+			reader := bufio.NewReader(resp.Body)
+			for {
+				// 0x0a is newline
+				n, err := reader.ReadBytes(0x0a)
+				if err == io.EOF {
+					break
+				} else if err != nil {
+					results <- subscraping.Result{Source: sourceName, Type: subscraping.Error, Error: err}
+					s.errors++
+					session.DiscardHTTPResponse(resp)
+					return
+				}
+
+				var response safResponse
+				err = jsoniter.Unmarshal(n, &response)
+				if err != nil {
+					results <- subscraping.Result{Source: sourceName, Type: subscraping.Error, Error: err}
+					s.errors++
+					session.DiscardHTTPResponse(resp)
+					return
+				}
+
+				// Cond is a scalar enum of string values: {“begin”, “ongoing”, “succeeded”, “limited”, “failed”}
+				// See https://www.domaintools.com/resources/user-guides/farsight-streaming-api-framing-protocol-documentation/
+				// for more details
+				respCond = response.Cond
+				if (respCond == "" || respCond == "ongoing") && response.Obj.Name != "" {
+					results <- subscraping.Result{
+						Source: s.Name(), Type: subscraping.Subdomain, Value: strings.TrimSuffix(response.Obj.Name, "."),
+					}
+					s.results++
+				} else if respCond == "failed" {
+					// DNSDB's terminating jsonl object's cond is "failed", notify the user.
+					err = fmt.Errorf("%s terminated with condition: %s", sourceName, respCond)
+					results <- subscraping.Result{Source: sourceName, Type: subscraping.Error, Error: err}
+					s.errors++
+					session.DiscardHTTPResponse(resp)
+					return
+				} else if respCond == "limited" && rate.OffsetMax.String() != rateNA {
+					if s.results > offsetMax {
+						// Can't offset any higher so all done.
+						break
+					}
+					// Set respCondition to "" to get more results with the offset set to s.results
+					queryParams.Set("offset", strconv.FormatUint(s.results, 10))
+					respCond = ""
+				}
 			}
-			s.results++
+			session.DiscardHTTPResponse(resp)
```

Comment:
should have no different effect from resp.Body.Close() ?

Reply:
Yep, changed